Implement new Codec that uses `encoding.BufferSlice` instead of `[]byte`

PapaCharlie opened the pull request:
This new codec, as outlined in https://github.com/grpc/grpc-go/issues/6619 will allow the reuse of buffers to their fullest extent. Note that this deliberately does not (yet) implement support for stathandlers, however all the relevant APIs have been updated and both old and new Codec implementations are supported.

PapaCharlie committed:
Implement new Codec that uses [][]byte instead of []byte

This new codec, as outlined in https://github.com/grpc/grpc-go/issues/6619 will allow the reuse of buffers to their fullest extent. Note that this deliberately does not (yet) implement support for stathandlers, however all the relevant APIs have been updated and both old and new Codec implementations are supported.

PapaCharlie committed:
Disable full materialization if the BufferSlice only has one buffer in it

PapaCharlie committed:
Fix tests

PapaCharlie committed:
Add copyright

PapaCharlie committed:
Complete migration by reading off wire in *Buffers

PapaCharlie committed:
More fixes

PapaCharlie committed:
transport package mostly fixed!

PapaCharlie committed:
Nuke `BufferProvider` interface

PapaCharlie committed:
Move to new `mem` package, wire in the BufferPool all the way down to the transport package

PapaCharlie committed:
Complete!

PapaCharlie committed:
Address vet

PapaCharlie committed:
Merge remote-tracking branch 'upstream/master' into pc/newcodec

PapaCharlie committed:
Always fully materialize the buffers for the binlog

PapaCharlie committed:
First pass at attempting getting the leakcheck package at checking buffers

PapaCharlie committed:
Implement buffer leak check

PapaCharlie committed:
More frees

PapaCharlie committed:
Make referencing a buffer actually create a copy

PapaCharlie committed:
Clear unread buffers on stream close

PapaCharlie committed:
Add ref in SendMsg

PapaCharlie committed:
Fix retries

PapaCharlie committed:
Fix the retries?

PapaCharlie committed:
Drain stream to pass test

PapaCharlie committed:
Cleanups

PapaCharlie committed:
Merge remote-tracking branch 'upstream/master' into pc/newcodec

PapaCharlie committed:
Remove dial

PapaCharlie committed:
Disable goroutine leak check, but never disable buffer leak check

PapaCharlie committed:
Merge remote-tracking branch 'upstream/master' into pc/newcodec

PapaCharlie committed:
Merge remote-tracking branch 'upstream/master' into pc/newcodec

easwars commented on mem/buffer_pool.go:
```diff
@@ -0,0 +1,156 @@
+}
+
+func (p *sizedBufferPool) Put(buf []byte) {
+	buf = buf[:cap(buf)]
+	// TODO: replace this loop with `clear`, though the compiler should be smart
+	// enough to optimize this.
+	for i := range buf {
+		buf[i] = 0
+	}
+	p.pool.Put(&buf)
```
Do we need this logic to zero fill buffers added to the pool? I'm interested to know if there is a real use case where you needed this behavior?

I believe the old buffer pool implementation explicitly said that it does guarantee that returned buffers would be zero filled.

PapaCharlie replied:
My reasoning for it was that there's no telling how long these buffers will stay in the pool. If something sensitive like passwords are in those buffers, it could be problematic to keep them around longer than they should be. Also, because `Buffer.ReadOnlyData` returns the raw slice, readers may index past the slice's `len` (if `len < cap`) and reveal the original data. Just seems like a good idea, especially now that 1.21 introduced `clear`

easwars replied:
Since we dropped support got Go1.20, we should be good to use `clear` here. Thanks.

PapaCharlie replied:
Yep, all done

easwars commented on mem/buffer_pool.go:
Are the changes here only an improvement to what we previously had in `shared_buffer_pool.go`? I don't see new functionality added here. Just confirming.

PapaCharlie replied:
There is an improvement. The old `shared_buffer_pool.go` only provided [fixed sized pools](https://github.com/grpc/grpc-go/blob/master/shared_buffer_pool.go#L52-L57). This new version allows the caller to specify what pool sizes to use. This makes it better optimized for gRPC's default max frame size of 16KB, which wasn't included in the original implementation.

easwars replied:
Thanks.

easwars commented on mem/buffers.go:
```diff
@@ -0,0 +1,217 @@
+	return b
+}
+
+func Copy(data []byte, pool BufferPool) *Buffer {
```
This is more like `NewBufferFromPool`, right? Along with a copy, so the caller can make changes to the byte slice that they passed?

PapaCharlie replied:
Yeah, it wanted to make it unambiguous that there was a copy happening. This is, after all, motivated by the reduction of copies. I also didn't want it to be too wordy, so I picked `Copy`. Completely open to suggestions though.

easwars replied:
Thought of a few options, but none very convincing:
- Add a method to the `BufferPool` interface to make a `Buffer` given a `[]byte`
  - The disadvantage here is that it seems to pollute the interface which now aligns well with `sync.Pool`
- A `NewBufferFromPool` with the same signature as this one
  - The name is longer, but more descriptive, but doesn't quite capture the fact that the given `[]byte` is copied
- A `NewBufferFromPoolWithCopy` with the same signature as this one
  - Feels more Java than Go
  
Maybe when we add the documented, I might be convinced with `Copy`.

PapaCharlie replied:
This has been documented, want to take another look?

easwars commented on mem/buffers.go:
```diff
@@ -0,0 +1,217 @@
+	data  []byte
+	refs  *atomic.Int32
+	free  func([]byte)
+	freed bool
```
I see that `freed` is an indication of whether the `Buffer` is freed, and not whether the underlying byte slice is freed by invoking the `free` func. The latter happens only when all refs are released.

This also means that the `Buffer` type is not safe for concurrent access. Is that by design? If so, why?

PapaCharlie replied:
Each instance of `Buffer` is not intended to be safe for concurrent access. It made the code easier to read/write, and simplifies the uses a lot. For example, if these buffers were safe for concurrent access, you could end up in a situation where one goroutine frees the buffer while another is accessing it. Instead, making this explicitly not safe (still writing the documentation for that) discourages this pattern, and instead each goroutine should get its own reference and manage the lifecycle independently.

easwars replied:
Thanks for the explanation. I'm good to making the type unsafe for concurrent access, as long as document so.

PapaCharlie replied:
This has been documented 👍

easwars commented on mem/buffers.go:
```diff
@@ -0,0 +1,217 @@
+
+func (b *Buffer) ReadOnlyData() []byte {
+	if b.freed {
+		panic("Cannot read freed buffer")
```
Is the eventual plan to replace `panics` with errors returned from the API?

PapaCharlie replied:
Actually, no. The intent for this was to effectively make it behave like a nil pointer. Once it's been freed, it should behave like accessing invalid memory. What do you think?

easwars replied:
`panics` are never nice. I would prefer the APIs returning errors instead. They could also log the errors at `ERROR` verbosity, and when this happens in our tests, the tests fail (even if the actual function passed).

PapaCharlie replied:
They aren't nice but this feels like it's on the level of trying to write to a `nil` map or sending on a closed channel. There is no way to recover from it and it indicates a serious bug

easwars commented on mem/buffers.go:
```diff
@@ -0,0 +1,217 @@
+	return n, nil
+}
+
+func NewWriter(buffers *BufferSlice, pool BufferPool) *Writer {
```
Can `NewWriter` return an `io.Writer` and thereby we can unexport the `Writer` type?

PapaCharlie replied:
Sure, makes sense!

easwars commented on stream.go:
```diff
@@ -568,8 +569,13 @@ type clientStream struct {
 	onCommit   func()
-	buffer     []func(a *csAttempt) error // operations to replay on retry
-	bufferSize int                        // current size of buffer
+	buffer     []replayOp // operations to replay on retry
+	bufferSize int        // current size of buffer
```
Much of this code was reading way too long ago, and has variable names which are way too short. Given that we have other buffers, could we name these `replayBuffer` and `replayBufferSize`?

PapaCharlie replied:
Yeah makes sense, can do

easwars commented on rpc_util.go:
```diff
@@ -586,14 +586,14 @@ type parser struct {
 	// See the comment on recvMsg for the permissible
 	// error types.
-	r io.Reader
+	r transport.Reader
```
Could we define this interface here. It could be unexported interface as well.

Go interfaces generally belong in the package that uses values of the interface type, not the package that implements those values

PapaCharlie replied:
`transport` is an internal package, so it's not actually public. But I can move it to `rpc_util` and make it private, especially since that's the only place it's actually used.

PapaCharlie replied:
Should I declare it like this:
```go
// parser reads complete gRPC messages from the underlying reader.
type parser struct {
	// r is the underlying reader.
	// See the comment on recvMsg for the permissible
	// error types.
	r interface {
		Read(n int) (mem.BufferSlice, error)
	}

	// The header of a gRPC message. Find more detail at
	// https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md
	header [5]byte

	// bufferPool is the pool of shared receive buffers.
	bufferPool mem.BufferPool
}
```
Or should I actually declare a standalone type, like this?
```go
type bufferReader interface{
	Read(n int) (mem.BufferSlice, error)
}

// parser reads complete gRPC messages from the underlying reader.
type parser struct {
	// r is the underlying reader.
	// See the comment on recvMsg for the permissible
	// error types.
	r bufferReader

	// The header of a gRPC message. Find more detail at
	// https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md
	header [5]byte

	// bufferPool is the pool of shared receive buffers.
	bufferPool mem.BufferPool
}
```
There are no other uses of this interface

easwars replied:
> `transport` is an internal package, so it's not actually public

But `rpc_util.go` is part of the top-level `grpc` package and has access to everything inside `internal`.

PapaCharlie replied:
Yeah but I meant that it's not actually public to users of `grpc`, it's only public internally. Either way, this has been moved

easwars reviewed (commented):
I've only hardly scratched the surface yet, but I just wanted to put in my thoughts so far as comments, so that we can discuss them once I'm ready.

Please expect more over the next few days, and then we can discuss. Thanks.

easwars commented on experimental/experimental.go:
```diff
@@ -28,23 +28,24 @@ package experimental
 // enabled.
-func WithRecvBufferPool(bufferPool grpc.SharedBufferPool) grpc.DialOption {
-	return internal.WithRecvBufferPool.(func(grpc.SharedBufferPool) grpc.DialOption)(bufferPool)
+func WithRecvBufferPool(bufferPool mem.BufferPool) grpc.DialOption {
```
Should we rename this to `WithBufferPool` since this buffer pool is now used for the send and receive sides?

PapaCharlie replied:
Good idea, that makes more sense

easwars commented on internal/transport/http2_client.go:
```diff
@@ -346,7 +347,7 @@ func newHTTP2Client(connectCtx, ctx context.Context, addr resolver.Address, opts
 		streamsQuotaAvailable: make(chan struct{}, 1),
 		keepaliveEnabled:      keepaliveEnabled,
-		bufferPool:            newBufferPool(),
+		bufferPool:            opts.BufferPool,
```
Earlier, there was one shared pool per transport. Now, there is one shared pool per clientConn. And this is going to be used for the send and receive sides. Does this change anything, like increased contention while `getting` and `putting` buffers from and to the pool?

PapaCharlie replied:
Hmmm that's an interesting question. I think there should be some benefits to sharing as much as possible but it's possible it could indeed increase contention. Probably something worth benchmarking?

easwars commented on internal/transport/transport.go:
```diff
@@ -296,6 +278,8 @@ type Stream struct {
 	// this must be lowercase or the behavior is undefined.
 	contentSubtype string
+
+	bufferPool mem.BufferPool
```
Do we need this field in the `Stream`?

I see that this is being set from three places:
- When creating a stream from a client transport
- When creating a stream from a server transport
- When creating a stream from a server handler transport

`Stream.bufferPool` is accessed from methods on the client transport or server transport, both named `handleData`, and in both places the same buffer pool is available in the transport as well.

Am I missing something?

PapaCharlie replied:
Yeah it was because the transports are defined like this:
```
	st           ServerTransport    // nil for client side Stream
	ct           *http2Client       // nil for server side Stream
```
It works for `ct` since it's a raw type but `st` is an interface which doesn't expose its underlying buffer pool. I did just update the interface to allow this and removed this field

easwars commented on mem/buffers.go:
```diff
@@ -0,0 +1,217 @@
+	b.refs.Add(1)
+
+	data := b.data
+	var free func([]byte)
+	if f := b.free; f != nil {
+		free = func(_ []byte) {
+			f(data)
+		}
+	}
```
- Lets say we have a `Buffer`, `b1` and is the only ref to the underlying `[]byte`
- Now, if we do `b2 := b1.Split(n)`, the refs on both `b1` and `b2` are going to `2`
- But `b1.free` is now `nil`
- Let's say `b2` is freed first, and it decrements `refs` and since it is not yet `0`, it wont call the free function
- And when `b1` is freed, the `refs` reach `0`, but the free function is `nil`. So, the underlying `[]byte` is not freed.

Am I missing something?

easwars commented on mem/buffers.go:
```diff
@@ -0,0 +1,217 @@
+	return n, nil
+}
+
+func NewWriter(buffers *BufferSlice, pool BufferPool) *Writer {
```
The point of passing a `BufferSlice` to the `Writer` is to eventually get to the written data, right?

Would a `Writer` method on a `BufferSlice` make sense, analogous to the `Reader` method, and the `Writer` method would take a `BufferPool` as argument.

PapaCharlie replied:
All the methods on `BufferSlice` are direct, instead of on `*BufferSlice`, e.g.:
```go
func (s BufferSlice) Materialize() []byte {
	out := make([]byte, s.Len())
	s.CopyTo(out)
	return out
}
```
In order for the `Writer` to be able to append to the `BufferSlice` it needs to be a pointer method. It would mean defining mixed pointer/non-pointer receiver methods, which is why I avoided it. The alternative would be to define _all_ `BufferSlice` methods on the pointer receiver, but it feels like that makes it less obvious that `BufferSlice` is meant to be immutable. I also do like the ergonomics of using `NewWriter` like this:
```go
var out mem.BufferSlice
w := mem.NewWriter(&out, pool)
```
vs this
```go
var out mem.BufferSlice
w := (&out).NewWriter(pool)
```
(^ the above example doesn't use `out := new(mem.BufferSlice)` because the codec API uses `mem.BufferSlice`, not `*mem.BufferSlice`)

HippoBaro commented:
Very happy to see this; I'll try to find some time to deploy this and report how it fares compared with https://github.com/grpc/grpc-go/pull/6613

easwars commented on mem/buffer_pool.go:
```diff
@@ -0,0 +1,156 @@
+// buffer pools for different sizes of buffers.
+type tieredBufferPool struct {
+	sizedPools   []*sizedBufferPool
+	fallbackPool *sizedBufferPool
```
I'm not completely convinced with using a `sizedBufferPool` for fallback. Could we instead use something like a `nopBufferPool` from the now deleted `shared_buffer_pool.go`. The reason I'm saying this is because:
- a fallback pool created as a `sizedBufferPool` of size `0`, will always return an empty slice from its `New` function, and `Get` will always add this back to the pool because `if cap(bs) < size` will always be `true`
- and if we don't expect the `sizedBufferPool` to work with a size of `0`, we can get rid of the extra code in `Get`, which can now simply be:
```golang
// Get assumes that size is always <= defaultSize
func (p *sizedBufferPool) Get(size int) []byte {
	bs := *p.pool.Get().(*[]byte)
	return bs[:size]
}
```

PapaCharlie replied:
You are right, there's a very good chance the current implementation of `Get` will initially dump a bunch of empty buffers in the pool. The old `nopBufferPool` is actually preserved here, but it's not clever enough IMO because it doesn't actually pool buffers, it just creates them on demand. I will add a new `simpleBufferPool` implementation that does on-demand stuff but still pools the buffers as best it can.

PapaCharlie replied:
Done

easwars commented on codec.go:
```diff
@@ -21,18 +21,75 @@ package grpc
+type codecV0Bridge struct {
+	codec interface {
+		Marshal(v any) ([]byte, error)
+		Unmarshal(data []byte, v any) error
+	}
+}
+
+func (c codecV0Bridge) Marshal(v any) (mem.BufferSlice, error) {
+	data, err := c.codec.Marshal(v)
+	if err != nil {
+		return nil, err
+	} else {
+		return mem.BufferSlice{mem.NewBuffer(data, nil)}, nil
+	}
```
Nit: The else is not required:

```golang
data, err := c.codec.Marshal(v)
if err != nil {
	return nil, err
}
return mem.BufferSlice{mem.NewBuffer(data, nil)}, nil
```

PapaCharlie replied:
Sure, fair enough

easwars commented on encoding/proto/proto_v2.go:
```diff
@@ -0,0 +1,72 @@
+// gRPC.
+type codecV2 struct{}
+
+var _ encoding.CodecV2 = (*codecV2)(nil)
```
Curious, does this line buy us anything at all, since the call to `encoding.RegisterCodecV2` will fail if the `encoding.CodecV2` undergoes a breaking change.

PapaCharlie replied:
No, it doesn't, but on principle I always add these. Can remove if you feel strongly

easwars replied:
That's fine.

easwars commented on encoding/proto/proto_v2.go:
```diff
@@ -0,0 +1,72 @@
+	if err != nil {
+		pool.Put(buf)
+		return nil, err
+	} else {
+		return mem.BufferSlice{mem.NewBuffer(buf, pool.Put)}, nil
```
Same here. We can do without the `else`.

PapaCharlie replied:
👍

easwars commented on encoding/proto/proto_v2.go:
```diff
@@ -0,0 +1,72 @@
+		return nil, fmt.Errorf("proto: failed to marshal, message is %T, want proto.Message", v)
+	}
+
+	pool := mem.DefaultBufferPool()
```
Is there a way to get the encoding plugins to use the buffer pool configured on the channel, instead of using the default buffer pool here? One obvious way would be to pass the buffer pool in calls to `Marshal` and `Unmarshal`, but feels clunky.

PapaCharlie replied:
Yeah that definitely feels clunky. Technically however, it's still possible when it's done the other way around. When creating the channel, the caller can pass in the pool with `WithBufferPool` and a `CodecV2` implementation that uses the same pool. It's equally clunky but at least leaves the door open to that. The idea here is that the default pool should be good enough for 99% of use cases, and extreme power users will be fine jumping through hoops for their performance gains from custom pools. Thoughts?

PapaCharlie replied:
Remember that the default marshaller outputs everything to a single large buffer instead of smaller chunked buffers that play better with buffer pools. I think anyone that requires their custom pools will likely also provide custom codecs.

easwars replied:
Thanks for the explanation.

easwars commented on stream.go:
```diff
@@ -891,23 +902,44 @@ func (cs *clientStream) SendMsg(m any) (err error) {
+		return status.Errorf(codes.ResourceExhausted, "trying to send message larger than max (%d vs. %d)", payloadLen, *cs.callInfo.maxSendMessageSize)
 	}
+
+	// always take an extra ref in case data == payload (i.e. when the data isn't
```
The logic here isn't super clear to me. Could you please check my understanding here?

- The deferred calls to `Free` on `data` and `payload` release the existing references acquired as part of `prepareMsg`
- There is one unconditional reference to payload that is taken in this method. How/when is this released?
  - When the `loopWriter` actually writes the data to the framer?
- There is one reference to `payloadRef` taken everytime `a.sendMsg` is called. This happens for the first send and any subsequent retries. How/when are these references released?
  - If the `op` is not queued for retry, it is released in the check `if !onSuccessCalled`
  - if the `op` is queued for retry, it is released as part of `replayOp.cleanup`
  
Comments:
- Can we be a little more descriptive in the comments here about when exactly every ref is freed?
- Can we get rid of the `payloadRef` variable and instead call `payload.Ref` in the call to `sendMsg` and everywhere where we call `payloadRef.Free` in this func, it would become `payload.Free`

PapaCharlie replied:
This was a nightmare to debug and figure out. The logic is pretty complicated and annoying to reason about. I can try to improve the comment, and let me look at whether we can remove the additional ref here

easwars commented on stream.go:
```diff
@@ -1101,17 +1137,17 @@ func (a *csAttempt) recvMsg(m any, payInfo *payloadInfo) (err error) {
-			CompressedLength: payInfo.compressedLength,
-			Length:           len(payInfo.uncompressedBytes),
-		})
+	if len(a.statsHandlers) != 0 {
```
Why do we need this `len` check?

PapaCharlie replied:
Ah, we used to! Because it used to materialize the buffer to go into the `Data` field. However, now that that field has been removed, I can also remove that `if`

easwars commented on xds/internal/xdsclient/authority_test.go:
```diff
@@ -176,7 +176,7 @@ func (s) TestTimerAndWatchStateOnErrorCallback(t *testing.T) {
 //   - update the mgmt server with resourceB and verifies that the watcher's
 //     update callback is invoked.
-func (s) TestWatchResourceTimerCanRestartOnIgnoredADSRecvError(t *testing.T) {
+func TestWatchResourceTimerCanRestartOnIgnoredADSRecvError(t *testing.T) {
```
Hmm ... interesting, why do we need this?

PapaCharlie replied:
Just an error, I brought it back

easwars commented on server.go:
```diff
@@ -1347,24 +1362,25 @@ func (s *Server) processUnaryRPC(ctx context.Context, t transport.ServerTranspor
-				Data:             d,
-			})
+
+		if len(shs) != 0 {
```
These `len` checks only seem to increase the indent of the code inside the `for` loop.

PapaCharlie replied:
Same as https://github.com/PapaCharlie/grpc-go/pull/1#discussion_r1655395873, this used to be necessary, now it's not

PapaCharlie committed:
Add docs and address comments

PapaCharlie committed:
Getting closer

PapaCharlie committed:
Disable leakcheck

PapaCharlie committed:
Add TODO on leakcheck

PapaCharlie committed:
Address comments

PapaCharlie committed:
Address more comments

PapaCharlie committed:
Clean up buffers when initial call succeeds

PapaCharlie committed:
More touchups

easwars commented on codec.go:
```diff
@@ -21,18 +21,79 @@ package grpc
-var _ baseCodec = Codec(nil)
-var _ baseCodec = encoding.Codec(nil)
+func newCodecV0Bridge(c Codec) baseCodec {
+	return codecV0Bridge{c}
```
Nit: here and in `newCodecV1Bridge`, could we specify the field name in the literal struct initialization?

PapaCharlie replied:
Sounds good, done

easwars commented on experimental/experimental.go:
```diff
@@ -28,38 +28,37 @@ package experimental
-// the application's workload, this could result in reduced memory allocation.
+// WithBufferPool returns a grpc.DialOption that configures the use of bufferPool
+// for parsing incoming messages on a grpc.ClientConn, and for temporary buffers
+// when marshaling outgoing messages. By default, mem.DefaultBufferPool is used,
```
Would it be better off to start with the usage of the buffer pool being turned off by default, and having the users explicitly set this dial/server option to enable it?

If we think that it is better to start with the default being enabled, we need to clearly document how to disable it.

easwars commented on experimental/experimental.go:
```diff
@@ -28,38 +28,37 @@ package experimental
-	return internal.WithRecvBufferPool.(func(grpc.SharedBufferPool) grpc.DialOption)(bufferPool)
+// Note: The following options will interfere with the buffer pool because they
+// require a fully materialized buffer instead of a sequence of buffers:
+// WithStatsHandler, EnableTracing, and binary logging. In such cases,
```
The `WithStatsHandler` could be used without the user being aware of it being used. For example, if the user turns on opentelemetry metrics/tracing, the implementation under the hood uses stats handlers.

PapaCharlie replied:
As mentioned on the call, this is actually not the case anymore, will update the docs.

easwars commented on codec.go:
```diff
@@ -21,18 +21,79 @@ package grpc
+	if err != nil {
+		return nil, err
+	}
+	return mem.BufferSlice{mem.NewBuffer(data, nil)}, nil
```
Does it make sense to use the default pool here? I guess the downside there is that we would have to make a copy. But here the `data` slice would only be freed after all refs are released and the GC runs, right? What are the tradeoffs here? Thanks.

PapaCharlie replied:
It's not really possible since `c.codec.Marshal` already created the buffer and didn't actually reuse anything. The best that this would do is add a buffer to the pool once it's done. Since (in theory) all buffers that leave the pool eventually return there, adding additional buffers shouldn't actually provide any benefit.

easwars commented on encoding/proto/proto_v2.go:
```diff
@@ -0,0 +1,71 @@
+
+	pool := mem.DefaultBufferPool()
+	buf := pool.Get(proto.Size(vv))
+	_, err := proto.MarshalOptions{}.MarshalAppend(buf[:0], vv)
+	if err != nil {
```
Nit: this can be on a single line.

PapaCharlie replied:
Interestingly, this is actually invalid syntax:
```go
	if _, err := proto.MarshalOptions{}.MarshalAppend(buf[:0], vv); err != nil {
		pool.Put(buf)
		return nil, err
	}
```
```
% go build ./...
# google.golang.org/grpc/encoding/proto
encoding/proto/proto_v2.go:47:12: syntax error: cannot use _, err := proto.MarshalOptions as value
encoding/proto/proto_v2.go:48:16: syntax error: unexpected newline in composite literal; possibly missing comma or }
encoding/proto/proto_v2.go:49:3: syntax error: unexpected return at end of statement
encoding/proto/proto_v2.go:51:2: syntax error: non-declaration statement outside function body
```
Which is fascinating. It's parsing the `{` after `proto.MarshalOptions` as the start of the `if` block. Anyway, fixed with:
```go
	if _, err := (proto.MarshalOptions{}).MarshalAppend(buf[:0], vv); err != nil {
		pool.Put(buf)
		return nil, err
	}
```
Just interesting

easwars commented on encoding/proto/proto_v2.go:
```diff
@@ -0,0 +1,71 @@
+
+	buf := data.LazyMaterialize(mem.DefaultBufferPool())
+	defer buf.Free()
+	// TODO: Upgrade proto.Unmarshal to support encoding.BufferSlice
```
Is there a known plan for how to do this? If so, can we capture it here in the TODO.

Also, this should be `mem.BufferSlice` and not `encoding.BufferSlice`.

easwars commented on mem/buffer_pool.go:
```diff
@@ -0,0 +1,177 @@
+	return *defaultBufferPool.Load()
+}
+
+func SetDefaultBufferPool(pool BufferPool) {
```
Is this function meant to be used only for testing? If so, can we add a `ForTesting` suffix to the function name. Also,  we need exported types in this file to have docstrings. Thanks.

PapaCharlie replied:
In this repo, it's only used for testing. I guess in practice it's probably better to expect that people that need to provide a custom pool do so with the call options but it's probably more convenient to let them override the default one? I can make it `ForTesting` for now and we can see if it's something that ends up actually being called outside of testing purposes.

easwars replied:
Yes, +1 for `ForTesting` for now. Thanks.

PapaCharlie replied:
Sounds good

easwars commented on preloader.go:
```diff
@@ -58,10 +59,14 @@ func (p *PreparedMsg) Encode(s Stream, msg any) error {
 	}
 	p.encodedData = data
-	compData, err := compress(data, rpcInfo.preloaderInfo.cp, rpcInfo.preloaderInfo.comp)
+	// TODO: it should be possible to grab the bufferPool from the underlying
+	// stream implementation.
```
How would one do that? It would be nice if the TODO was actionable. Thanks.

PapaCharlie replied:
Done

easwars commented on internal/grpctest/grpctest.go:
```diff
@@ -48,16 +49,22 @@ type Tester struct{}
 // Setup updates the tlogger.
 func (Tester) Setup(t *testing.T) {
 	TLogger.Update(t)
+	// TODO: There is one final leak around closing connections without completely
```
Is this TODO already addresses? I don't see any commented out code here.

PapaCharlie replied:
So we talked about this in the meeting, I will add a per-test leakcheck override instead of globally disabling it.

easwars commented on internal/leakcheck/leakcheck.go:
```diff
@@ -16,18 +16,81 @@
-// Package leakcheck contains functions to check leaked goroutines.
+// Package leakcheck contains functions to check leaked goroutines and buffers.
 //
 // Call "defer leakcheck.Check(t)" at the beginning of tests.
```
There is no `Check` method anymore?

PapaCharlie replied:
Yeah you are right. I have updated this and provided a better method

easwars commented on internal/grpctest/grpctest.go:
```diff
@@ -48,16 +49,22 @@ type Tester struct{}
 
 // Teardown performs a leak check.
 func (Tester) Teardown(t *testing.T) {
+	leakcheck.CheckTrackingBufferPool()
```
Does anything change in terms of tests passing/failing, if we perform this check after the check for goroutines? Or was there a specific reason to perform this check first? If so, could we have a comment for that. Thanks.

PapaCharlie replied:
Probably not? I can see if it does

easwars commented on mem/buffers.go:
```diff
@@ -0,0 +1,288 @@
+		}
+	}
+
+	b.data = data[:n]
```
This changes the receiver to point to the first `n` bytes, but the docstring says otherwise.

PapaCharlie replied:
Oops, updated the docs

easwars commented on server.go:
```diff
@@ -342,7 +343,22 @@ func CustomCodec(codec Codec) ServerOption {
+// ForceServerCodecV2 is the equivalent of ForceServerCodec, but for the new
+// CodecV2 interface.
+//
+// Will be supported throughout 1.x.
```
We generally add this line when we deprecate APIs I thought.

PapaCharlie replied:
Yeah I should remove that

PapaCharlie replied:
There are other functions in this file that have the same comment (for example [this one](https://pkg.go.dev/google.golang.org/grpc#ForceServerCodec))

easwars commented on stats/stats_test.go:
```diff
@@ -538,40 +538,29 @@ func checkInPayload(t *testing.T, d *gotData, e *expectedData) {
 	}
+
+	expectedPayload := payloads[*idx]
+	if !proto.Equal(st.Payload.(proto.Message), expectedPayload) {
```
Could we use `cmp.Diff` instead, and pass it a ` protocmp.Transform()` option?

Something like:
```golang
if diff := cmp.Diff(want, got, protocmp.Transform()); diff != "" {
    t.Fatalf("unexpected difference in protobuf messages (-want +got):\n%s", diff)
}
```

PapaCharlie replied:
Done

easwars commented on stats/stats_test.go:
```diff
@@ -538,40 +538,29 @@ func checkInPayload(t *testing.T, d *gotData, e *expectedData) {
+		payloads = e.requests
 	}
+
+	expectedPayload := payloads[*idx]
```
Nit: s/expectedPayload/wantPayload/

PapaCharlie replied:
Done

easwars commented on stats/stats_test.go:
```diff
@@ -538,40 +538,29 @@ func checkInPayload(t *testing.T, d *gotData, e *expectedData) {
+		t.Fatalf("st.Payload = %v, want %v", st.Payload, expectedPayload)
+	}
+	*idx++
+	if st.Length != proto.Size(expectedPayload) {
+		t.Fatalf("st.Length = %v, want %v", st.Length, proto.Size(expectedPayload))
+	}
```
Maybe if we pass the whole struct to `cmp.Diff`, we wont have to check separate for individual fields here?

PapaCharlie replied:
I think there are fields we don't actually care about such as `RecvTime`. The original test was set up like this

easwars commented on stats/stats_test.go:
```diff
@@ -657,40 +646,29 @@ func checkOutPayload(t *testing.T, d *gotData, e *expectedData) {
-	// Below are sanity checks that WireLength and SentTime are populated.
+
+	expectedPayload := payloads[*idx]
+	if !proto.Equal(st.Payload.(proto.Message), expectedPayload) {
+		t.Fatalf("st.Payload = %v, want %v", st.Payload, expectedPayload)
+	}
+	*idx++
+	if st.Length != proto.Size(expectedPayload) {
+		t.Fatalf("st.Length = %v, want %v", st.Length, proto.Size(expectedPayload))
+	}
+
+	// Below are sanity checks that Length, CompressedLength and SentTime are populated.
 	// TODO: check values of WireLength and SentTime.
-	if len(st.Data) > 0 && st.WireLength == 0 {
+	if st.Length > 0 && st.WireLength == 0 {
 		t.Fatalf("st.WireLength = %v with non-empty data, want <non-zero>",
```
Similar comments as above.

PapaCharlie commented:
Closing to avoid confusion with https://github.com/grpc/grpc-go/pull/7356

PapaCharlie committed:
Address comments

PapaCharlie committed:
Allow specific tests to skip leakcheck

PapaCharlie committed:
Merge remote-tracking branch 'upstream/master' into pc/newcodec

PapaCharlie committed:
Introduce `mem` package

This package is required for https://github.com/grpc/grpc-go/pull/7356 but is
introduced early here to allow additional features to be built on top of it. It
provides a mechanism to reuse and recycle buffers to reduce GC pressure.

PapaCharlie committed:
Add ForceCodecV2 option

PapaCharlie committed:
Address comments

easwars committed:
add a few unit tests for Buffer; fix docstring for NopBufferPool

PapaCharlie committed:
Merge remote-tracking branch 'upstream/master' into pc/newcodec

PapaCharlie committed:
Make BufferSlice.Reader take reference

PapaCharlie committed:
Minor tweaks to BufferSlice.Reader

PapaCharlie committed:
Improve test

easwars committed:
more tests and minor touchups

- add more tests
- use got before want in tests
- cleanup some docstrings
- wrap comments at 80-cols

easwars committed:
fix a bug in BufferSlice.Reader

- and add tests for the BufferSlice

easwars committed:
add prefix to test names to match type

easwars committed:
handle review comments

- fix grammar in test error string
- move SetDefaultBufferPoolForTesting to the `internal` package

easwars committed:
handle more review comments from dfawley

PapaCharlie committed:
Merge remote-tracking branch 'upstream/master' into pc/mem

PapaCharlie committed:
Address comments

PapaCharlie committed:
Merge branch 'pc/mem' into pc/newcodec

PapaCharlie committed:
Update after rebase on pc/mem

PapaCharlie committed:
Rename test

PapaCharlie committed:
Merge branch 'pc/mem' into pc/newcodec

PapaCharlie committed:
Rename to onFree

PapaCharlie committed:
Merge branch 'pc/mem' into pc/newcodec

PapaCharlie committed:
Fix mem test

PapaCharlie committed:
Change to MaterializeToBuffer

PapaCharlie committed:
Merge branch 'pc/mem' into pc/newcodec

PapaCharlie committed:
Fix vet failure

PapaCharlie committed:
Unbreak uint cast

PapaCharlie committed:
Merge remote-tracking branch 'upstream/master' into pc/newcodec

PapaCharlie committed:
Reuse `Buffer` instances

PapaCharlie committed:
Fix `prepareMsg`

PapaCharlie committed:
Make `BufferPool.Put` take a `*[]byte`

PapaCharlie committed:
Hack hack hack!

PapaCharlie committed:
Implement flate.Reader to prevent wrapping

PapaCharlie committed:
Refactor splitting

PapaCharlie committed:
Run it with tags instead

PapaCharlie committed:
More touchups, tests are passing

PapaCharlie committed:
Make leaks informational instead of failures

PapaCharlie committed:
Enable test failures with tag

PapaCharlie committed:
Split more stuff

PapaCharlie committed:
Merge remote-tracking branch 'upstream/master' into pc/newcodec

PapaCharlie committed:
First pass at object reduction

PapaCharlie committed:
Getting closer

PapaCharlie committed:
Pool ref counters

PapaCharlie committed:
More fixes

PapaCharlie committed:
No-copy BufferSlice

PapaCharlie committed:
Finish up the job!

PapaCharlie committed:
Fix compile error on default tag

PapaCharlie committed:
Merge remote-tracking branch 'upstream/master' into pc/newcodec

PapaCharlie committed:
Merge remote-tracking branch 'upstream/master' into pc/newcodec

PapaCharlie committed:
Remove tags

PapaCharlie committed:
Merge remote-tracking branch 'upstream/master' into pc/newcodec

PapaCharlie committed:
Fix leakcheck

PapaCharlie committed:
Fix grpchttp2

easwars committed:
test/tools: update staticcheck version to latest (#7509)

PapaCharlie committed:
Fix vet

zasweq committed:
balancer/rls: Fix RLS Cache metrics (#7511)

PapaCharlie committed:
Fix tests

easwars committed:
grpclog: refactor to move implementation to grpclog/internal (#7465)

printchard committed:
transport: change `*http2Client` to interface `ClientTransport` (#7512)

easwars committed:
csds: unskip e2e test (#7502)

aranjans committed:
rls: update picker synchronously on configuration update (#7412)

arjan-bal committed:
Change version to 1.67.0-dev (#7520)

aranjans committed:
transport: add timeout for writing GOAWAY on http2Client.Close() (#7371)

dfawley committed:
rbac: fix usage of AuthInfo (#7522)

arjan-bal committed:
deps: update dependencies for all modules (#7531)

* Bump dependencies after release branch cut

* Regenerate protos

jhalterman committed:
Use builtin min and max functions (#7478)

* Use builtin min and max functions

Go added builtin min and max functions in 1.21. This commit removes existing functions and uses the built-ins in stead.

* Revert gofmt changes

menghanl committed:
doc: update keepalive ClientParameters doc about doubling the interval upon GOAWAY (#7469)

atollena committed:
ringhash: fix bug where ring hash can be stuck in transient failure despite having available endpoints (#7364)

PapaCharlie committed:
Merge remote-tracking branch 'upstream/master' into pc/newcodec

easwars committed:
bootstrap: add `String` method to ServerConfigs type (#7537)

PapaCharlie committed:
Address Doug's comments

PapaCharlie committed:
rerun tests

PapaCharlie committed:
Add coments

PapaCharlie committed:
Fix grammar

PapaCharlie committed:
Fix test

PapaCharlie committed:
Merge remote-tracking branch 'upstream/master' into pc/newcodec
